### PR TITLE
add windows build support

### PR DESCRIPTION
**Pull request by Leandros**
Added a Visual Studio solution. Confirmed working on Windows 11, with NeoVim Nightly.

### CODE REVIEW

**Comment by Conni2461** [comment]
I had something else in mind. I am not interested to merge a configuration for a proprietary system specific build system. Especially when a build with gcc (mingw) works on windows.
I need to look into how nvim-treesitter is handling windows builds and I honestly prefer a download script with a build in CI.

File: src/dllmain.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// dllmain.cpp : Defines the entry point for the DLL application.
+#define NOMINMAX
+#include <Windows.h>
+
+BOOL APIENTRY DllMain( HMODULE hModule,
```
Comment (Conni2461) [inline src/dllmain.cpp:5]: according to docs `DllMain` is an optional entry point: https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain

None the less no cpp files

File: src/fzf.h [hunk 2]
```diff
@@ -5,6 +5,12 @@
 #include <stdint.h>
 #include <stddef.h>
 
+#ifdef _WIN32
+    #define DLLEXPORT __declspec(dllexport)
+#else
+    #define DLLEXPORT
+#endif
```
Comment (Conni2461) [inline src/fzf.h:12]: Is `__declspec(dllexport)` a vs compiler thing ?

Comment (Leandros) [reply src/fzf.h:12]: It's a Windows thing. By default the visibility of all symbols is hidden (same thing can be achieved on gcc/clang with `-fvisibility=hidden`).

File: src/dllmain.cpp (hunk 1 above)
Comment (Leandros) [reply src/dllmain.cpp:5]: Yes, this wasn't supposed to be included, my mistake.

**Comment by Leandros** [comment]
> Edit: I would consider a build system generator like cmake or premake over a visual studio solution. Like how can you integrate it in vim. that it autobuilds on a PlugUpdate or PackerSync. If i push a commit and it requires a rebuild otherwise it the ffi glue code errors and i cant do this with a vs solution than its bad

You can do this with a vs solution using msbuild. That was the idea behind that.

**Comment by Conni2461** [comment]
Still requires msbuild. I thought about it some more and I dont want any windows only build files here. For compiler 1 c file.

You can just call the visual c compiler without build system configuration: https://docs.microsoft.com/en-us/cpp/build/reference/compiler-command-line-syntax?view=msvc-170

I just wrote a super quick and dirty downloader script with artifacts: https://github.com/nvim-telescope/telescope-fzf-native.nvim/tree/downloader

I would merge the `DLLEXPORT` stuff tho. That should be good

**Comment by Leandros** [comment]
Make sure to also prompt the user to install the required runtime libraries.

**Comment by Conni2461** [comment]
Fine. Remove the cpp file, fix formatting and ill merge ...

Note: I absolutely despise this solution. It physically hurts even thinking about it. So please fix the remaining things fast before i change my mind and just drop windows support.

**Comment by Leandros** [comment]
I'm sorry, that wasn't my intention. 

The ideal solution is two-fold:

1. ### Have a working Windows build (of which I provided a possible solution).
Ideally, the Windows build is entirely done natively (i.e without Cygwin or MinGW) since the C code is already fully portable.

2. ### Download the resulting binary in the post-setup script of the Vim plugin (of which you have provided the start of a possible solution).
With the CI already setup, providing a binary via github releases shouldn't be a huge hassle anymore. 

You need to be happy with both 1. and 2. and feel comfortable maintaining it. 
Your idea of having a CMake build is probably the best going forward into the future (unfortunately, my CMake experience is rather limiting).

**Comment by Leandros** [comment]
I have updated the PR to remove the VS build and add cmake instead.

**Comment by qgymib** [comment]
Hi,
I am working on windows port too and just found this article, it seems more completeness than mine.

After review the cmake script, there might be some improvement can be made:
1. modern cmake should use target-based rules.
2. cmake script should cover original Makefile-based functions.

Here is modified cmake script:

```
# CMakeLists.txt
cmake_minimum_required(VERSION 3.2)
project(fzf)

# add source files
add_library(${PROJECT_NAME}
    SHARED
        "src/fzf.c")

# define target include search paths
target_include_directories(${PROJECT_NAME}
    PUBLIC
        ${CMAKE_CURRENT_SOURCE_DIR}/src)

function(FZF_SETUP_COMPILE_OPTIONS name)
    if (CMAKE_C_COMPILER_ID STREQUAL "MSVC")
        target_compile_options(${name} PRIVATE /W4 /WX)
    else ()
        # Recommand to add -Wall -Wextra, should enable after fix compile error
        #target_compile_options(${name} PRIVATE -Wall -Wextra -Werror -std=gnu99)
        target_compile_options(${name} PRIVATE -std=gnu99)
    endif ()
endfunction()

# add warning check
FZF_SETUP_COMPILE_OPTIONS(${PROJECT_NAME})

# add test
if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
    include(CTest)
endif()
if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME AND BUILD_TESTING)
    add_subdirectory(test)
endif()
```

```
# test/CMakeLists.txt
# test
add_executable(fzf_test
    "test.c")
FZF_SETUP_COMPILE_OPTIONS(fzf_test)
target_link_libraries(fzf_test PRIVATE fzf)
add_test(fzf_test $<TARGET_FILE:fzf_test>)

# benchmark
add_executable(fzf_benchmark
    "benchmark.c")
FZF_SETUP_COMPILE_OPTIONS(fzf_benchmark)
target_link_libraries(fzf_benchmark PRIVATE fzf)
add_custom_target(benchmark
    COMMAND $<TARGET_FILE:fzf_benchmark>)

# ntest
add_custom_target(ntest
    COMMAND nvim
            --headless
            --noplugin
            -u test/minrc.vim
            -c "PlenaryBustedDirectory test/ { minimal_init = './test/minrc.vim' }")

# clangdhappy
add_custom_target(clangdhappy
    COMMAND compiledb make)
```

**Comment by Conni2461** [comment]
Thank you very much :) 

For me it was enough to execute these steps
```
mkdir build
cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release
cmake --build build
```

and we still need to check if it works with plug and packer. Can we do: `{ 'do': 'mkdir build && cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release && cmake --build build' } or should we write small sh and bat scripts? Also we need to add some docs, most likely in the README. Another thing we could do is maybe add a wiki entry that helps windows users get started.

I dont think the cmake stuff should cover everything in make. Especially clangdhappy because that just calls a tool to generate compile_commands.json, cmake can just do that

File: src/fzf.c [hunk 3]
```diff
@@ -1048,7 +1048,7 @@ static fzf_result_t fzf_call_alg(fzf_term_t *term, bool normalize,
  * - always v2 alg
  * - bool extended always true (thats the whole point of this isn't it)
  */
-fzf_pattern_t *fzf_parse_pattern(fzf_case_types case_mode, bool normalize,
+DLLEXPORT fzf_pattern_t *fzf_parse_pattern(fzf_case_types case_mode, bool normalize,
```
Comment (Shatur) [inline src/fzf.c:1051]: You can get rid of this and use [WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) in CMake.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,22 @@
+cmake_minimum_required(VERSION 3.13)
+project(fzf C)
+
+set(CMAKE_INSTALL_PREFIX ".")
+
+if (MSVC)
+    set(CMAKE_SHARED_LIBRARY_PREFIX lib)
+    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
+    add_compile_options(/W4)
```
Comment (Shatur) [inline CMakeLists.txt:9]: I would suggest to call `add_library` first and then use `target_compile_definitions` and `target_compile_options`.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+else()
+    add_compile_options(-Wall -Werror)
+    set(CMAKE_C_STANDARD 99)
+    set(CMAKE_C_STANDARD_REQUIRED TRUE)
```
Comment (Shatur) [inline CMakeLists.txt:13]: You would define this on top for both compilers.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+endif()
+
+# To build the library:
+#   $ mkdir build && cd build
+#   $ cmake .. -DCMAKE_BUILD_TYPE=Release
+#   $ cmake --build . --config Release
+add_library(fzf SHARED src/fzf.c src/fzf.h)
```
Comment (Shatur) [inline CMakeLists.txt:20]:
```suggestion
add_library(${PROJECT_NAME} SHARED src/fzf.c src/fzf.h)
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+install(TARGETS fzf DESTINATION ${CMAKE_BINARY_DIR})
```
Comment (Shatur) [inline CMakeLists.txt:22]:
```suggestion
install(TARGETS ${PROJECT_NAME} DESTINATION ${CMAKE_BINARY_DIR})
```

**Review by Shatur — CHANGES_REQUESTED** [review]

**Comment by Shatur** [comment]
> Especially clangdhappy because that just calls a tool to generate compile_commands.json, cmake can just do that

Agree, there is [CMAKE_EXPORT_COMPILE_COMMANDS](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html).

But I would move other stuff to CMake to make it more portable across different platforms.

**Comment by Conni2461** [comment]
We just merged cmake support: https://github.com/nvim-telescope/telescope-fzf-native.nvim/commit/2330a7eac13f9147d6fe9ce955cb99b6c1a0face

**Pull request closed without merging**